bentoml/_internal/cli/click_utils.py

```diff
@@ -110,65 +103,85 @@ def wrapper(quiet, verbose, config, *args, **kwargs):
+                )
                 return return_value
             except BaseException as e:
-                track_properties["duration"] = time.time() - start_time
-                track_properties["error_type"] = type(e).__name__
-                track_properties["error_message"] = str(e)
-                track_properties["return_code"] = 1
+                cli_properties["duration"] = time.time() - start_time
+                cli_properties["error_type"] = type(e).__name__
+                cli_properties["error_message"] = str(e)
+                cli_properties["return_code"] = 1
                 if type(e) == KeyboardInterrupt:
-                    track_properties["return_code"] = 2
-                track(TRACK_CLI_EVENT_NAME, track_properties)
+                    cli_properties["return_code"] = 2
+                track(
+                    CLI_TRACK_EVENT_TYPE, process_pid, event_properties=cli_properties
+                )
                 raise
 
         return wrapper
 
     @staticmethod
-    def raise_click_exception(func, cmd_group, **kwargs):
+    def raise_click_exception(
+        func: t.Union["t.Callable[P, t.Any]", "ClickFunctionWrapper[t.Any]"],
+        cmd_group: click.Group,
+        **kwargs: t.Any,
+    ) -> "ClickFunctionWrapper[t.Any]":
         command_name = kwargs.get("name", func.__name__)
 
         @functools.wraps(func)
-        def wrapper(*args, **kwargs):
+        def wrapper(*args: "P.args", **kwargs: "P.kwargs") -> t.Any:
             try:
                 return func(*args, **kwargs)
-            except BentoMLException as e:
-                msg = f"{cmd_group.name} {command_name} failed: {str(e)}"
-                raise ClickException(click.style(msg, fg="red"))
+            except BentoMLException as err:
+                msg = f"[{cmd_group.name}] `{command_name}` failed: {str(err)}"
+                raise ClickException(click.style(msg, fg="red")) from err
 
-        return wrapper
+        return t.cast("ClickFunctionWrapper[t.Any]", wrapper)
 
-    def command(self, *args, **kwargs):
+    def command(
+        self, *args: t.Any, **kwargs: t.Any
+    ) -> "t.Callable[[t.Callable[P, t.Any]], click.Command]":
         if "context_settings" not in kwargs:
             kwargs["context_settings"] = {}
         kwargs["context_settings"]["max_content_width"] = 120
 
-        def wrapper(func):
-            # add common parameters to command
+        def wrapper(func: "t.Callable[P, t.Any]") -> click.Command:
+            # add common parameters to command.
             func = BentoMLCommandGroup.bentoml_common_params(func)
             # Send tracking events before command finish.
             func = BentoMLCommandGroup.bentoml_track_usage(func, self, **kwargs)
-            # If BentoMLException raise ClickException instead before exit
+            # If BentoMLException raise ClickException instead before exit.
+            # NOTE: Always call this function last for type checker to work.
```

Comment:
why is that?

Reply:
because of type casting

Reply:
click dynamically inject `__click_params__` into a function, so there aren't any class we can use for type hint, thus the wrapper on L22